Source/Core/VideoCommon/ShaderGenCommon.cpp

```diff
@@ -362,3 +363,89 @@ const char* GetInterpolationQualifier(bool msaa, bool ssaa, bool in_glsl_interfa
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_NONE = {};\n",
+             static_cast<u32>(AttenuationFunc::None));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_POINT = {};\n",
+             static_cast<u32>(AttenuationFunc::Spec));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_DIR = {};\n",
+             static_cast<u32>(AttenuationFunc::Dir));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_SPOT = {};\n",
+             static_cast<u32>(AttenuationFunc::Spot));
+
+  // CUSTOM_SHADER_LIGHTING_LIGHT_TYPE "enum" values
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_LIGHT_TYPE_ALPHA = 0;\n");
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_LIGHT_TYPE_COLOR = 1;\n");
+
+  out->Write("struct CustomShaderLightData\n");
+  out->Write("{{\n");
+  out->Write("\tfloat3 position;\n");
+  out->Write("\tfloat3 direction;\n");
+  out->Write("\tfloat3 color;\n");
+  out->Write("\tuint attenuation_type;\n");
+  out->Write("\tuint light_type;\n");
+  out->Write("\tfloat4 cosatt;\n");
+  out->Write("\tfloat4 distatt;\n");
+  out->Write("}};\n\n");
+
+  // CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE "enum" values
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_PREV = 0;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_COLOR = 1;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_TEX = 2;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_RAS = 3;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_KONST = 4;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_NUMERIC = 5;\n");
+  out->Write("const uint CUSTOM_SHADER_TEV_STAGE_INPUT_TYPE_UNUSED = 6;\n");
+
+  out->Write("struct CustomShaderTevStageInputColor\n");
+  out->Write("{{\n");
+  out->Write("\tuint input_type;\n");
+  out->Write("\tfloat3 value;\n");
+  out->Write("}};\n\n");
+
+  out->Write("struct CustomShaderTevStageInputAlpha\n");
+  out->Write("{{\n");
+  out->Write("\tuint input_type;\n");
+  out->Write("\tfloat value;\n");
+  out->Write("}};\n\n");
+
+  out->Write("struct CustomShaderTevStage\n");
+  out->Write("{{\n");
+  out->Write("\tCustomShaderTevStageInputColor[4] input_color;\n");
+  out->Write("\tCustomShaderTevStageInputAlpha[4] input_alpha;\n");
+  out->Write("\tuint texmap;\n");
+  out->Write("\tfloat4 output_color;\n");
+  out->Write("}};\n\n");
+
+  // Custom structure for data we pass to custom shader hooks
+  out->Write("struct CustomShaderData\n");
+  out->Write("{{\n");
+  out->Write("\tfloat3 position;\n");
+  out->Write("\tfloat3 normal;\n");
+  if (numtexgens == 0)
+  {
+    // Cheat so shaders compile
+    out->Write("\tfloat3[1] texcoord;\n");
+  }
+  else
+  {
+    out->Write("\tfloat3[{}] texcoord;\n", numtexgens);
+  }
+  out->Write("\tuint texcoord_count;\n");
+  out->Write("\tuint[8] texmap_to_texcoord_index;\n");
+  out->Write("\tCustomShaderLightData[32] lights;\n");
```

Comment:
Thanks.  Yeah, that might make more sense, though also makes it more painful for the user.  It's still not clear to me what an "alpha" light even is.  I guess it's game specific.

It's also not clear to me why there are separate "channels", again, guessing this is game specific in terms of usage.

The ideal scenario is that the user doesn't care about any of the game specific functionality and can just loop over the lights for the draw call, doing whatever custom light functionality they want.

I suppose that's still possible with that notation.  Will play around with it!

Reply:
Some of the Jasper videos help with this - for instance the long video on Mario Galaxy mentions that the backlighting effect is done by having a light at the camera that affects the channel 1 alpha, which is used to *decrease* the brightness of the model. It's a bit weird and doesn't physically map to a real light, but it works well in the game.

I *think* the original intent of the two channels was that one was used for diffuse lighting and the other for specular highlights, but I don't know if that's how it's generally used in practice.

Reply:
Still uncertain what the best approach here would be.  One of Jasper's video said Mario Galaxy only had 3 real "lights" which from what I've seen, didn't surprise me, but still that's so few!  I wonder if channel 0 color is the typical light channel or if it really varies based on game.

Reply:
The main thing is, it feels a little cumbersome to have to loop over both color channels and both alpha channels.  But I also don't know if it makes sense to include them.  I suppose I'm fine going with all channels split.  Really won't know what is best until we get into it.

Reply:
There are eight hardware lights, and in effect, four lighting channels, COLOR0, ALPHA0, COLOR1, ALPHA1. ALPHA0/ALPHA1 can be configured entirely differently from COLOR0/COLOR1 -- entirely different light masks can be used, if wanted, the only restriction is that ALPHA0/ALPHA1 are a single value instead of tricolor, and COLOR/ALPHA are bundled together in a TEV stage when selected.

If you want to, you can think of alpha lights as just ways of doing a per-vertex dot product against an arbitrary other vector. That could be used to emulate a fresnel effect if modulated in the material -- that's exactly what Super Mario Galaxy is doing, for instance.